examples/with-react-intl/components/PageWithIntl.js

```diff
@@ -0,0 +1,44 @@
+import React, {Component} from 'react'
+import {IntlProvider, addLocaleData, injectIntl} from 'react-intl'
+
+// Register React Intl's locale data for the user's locale in the browser. This
+// locale data was added to the page by `pages/_document.js`. This only happens
+// once, on initial page load in the browser.
+if (typeof window !== 'undefined' && window.ReactIntlLocaleData) {
```

Comment:
> IMHO worth to create a bug on Next.js

There's already a bug and discussion for this.

> the localeDataScript itself (loaded by require.resolve + readFileSync) is just a string, which is serializable?

I want to avoid using `eval` in the browser because of CSP. You can see from the discussions about Next and CSP that ideally Next moves away from direct `eval`. So I don't want to introduce another place where `eval` is used.

> Right now the intl code is spread all around and that's hardly maintainable.

Can you be specific about what's not maintainable?

> should work by returning a serverRendered: !process.browser in your getInitialProps and wrap the addLocaleData in if (props.serverRendered)?:

This can work but the conditional should be `if (props.serverRendered && process.browser)` since there's no reason to call `addLocaleData()` in Node. The reason this API exists is that the locale data for all locales is too large to ship to the browser, but in Node the data already exists and is loaded.

----

I think an ideal change in Next is to use `serialize-javascript` instead of raw `JSON.stringify()` to serialize the result of calling the page's `getInitialProps()`. This would all the server to simply `require('react-intl/locale-data/' + lang)` and stick the locale data object on the `req`. This would allow the server rendered page to return `req.localeData` from its `getInitialProps()` which can be properly serialized by `serialize-javascript` and applied how you show in 3) in the browser.

All this said, it would be great to actually get this PR merged and continue making it better over time and as Next improves. If you feel that it's not maintainable and doesn't belong in this repo, then we can close this PR and I can have the official Next + React Intl example app live in React Intl's repo.